### ISSUE TITLE
1 ALPN per version.

### ISSUE THREAD

**Issue by martinduke** [quicwg/ops-drafts#360]
Applicability Sec 8:

>Note that HTTP/3's ALPN token ("h3") identifies not only the version of the application protocol, but also the version of QUIC >itself; this approach allows unambiguous agreement between the endpoints on the protocol stack in use.

I was going to start a discussion about this at IETF 111. I think we should think hard about this; if there are a lot of versions with small differences to the application, this could result in a ridiculous cloud of RFCs.

**Comment by britram** [comment]
Is there a change to make to this document in WGLC, or is this a note to be careful in the future?

**Comment by martinduke** [comment]
I disagree that the text of the H3 draft mandates this:

`HTTP/3 relies on QUIC version 1 as the underlying transport. The use of other QUIC transport versions with HTTP/3 MAY be defined by future specifications.`

That is a long way from saying "roll a new ALPN for each QUIC version".

Maybe everyone tacitly agrees we should roll a new ALPN, and I'm in the rough. But I'd like to confirm that.

**Comment by mirjak** [comment]
See quicwg/base-drafts#3117

**Comment by LPardue** [comment]
How about rewording the sentence to reflect reality, roughly

Since QUIC v1 (RFC9000) deferred defining a version negotiation mechanism. HTTP/3 (RFC-to-be-whatever) deferred deciding how it would map to other QUIC versions. HTTP/3 requires QUIC v1 and defines the ALPN token ("h3") to only apply to that version. This solves a versioning problem at the cost of future flexibility. Application protocol mappings written for QUIC v1 or other QUIC versions may be able to benefit from version negotiation mechanisms that did not exist during the development of HTTP/3 (RFC-to-be-whatever). Coupling QUIC version to ALPN tokens is one approach but not the only approach.

**Comment by mirjak** [comment]
I believe the text already says that this doesn't have to be applied by other application. The cited text is really only meant to described what was decided from h3 as an example. We can definitely reword this for reflect the consensus more clearly as proposed by Lucas.

**Comment by martinduke** [comment]
Lucas,

I would welcome your proposal as kicking the can down the road and not enshrining an approach that I think is bad.

I would also like to continue the ALPN/versioning discussion, but it doesn't have to block the applicability draft.
